akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -291,18 +332,75 @@ private[akka] class Shard(
     case msg: ShardQuery                         => receiveShardQuery(msg)
     case PassivateIdleTick                       => passivateIdleEntities()
     case msg: LeaseLost                          => receiveLeaseLost(msg)
-    case msg if extractEntityId.isDefinedAt(msg) => deliverMessage(msg, sender())
+    case msg: RememberEntityStoreCrashed         => rememberEntityStoreCrashed(msg)
+    case msg if extractEntityId.isDefinedAt(msg) => deliverMessage(msg, sender(), OptionVal.None)
+  }
+
+  def waitForAsyncWrite(entityId: EntityId, command: RememberEntitiesShardStore.Command)(
+      whenDone: EntityId => Unit): Unit = {
+    rememberEntitiesStore match {
+      case None =>
+        whenDone(entityId)
+
+      case Some(store) =>
+        if (VerboseDebug) log.debug("Waiting for async write of [{}] [{}]", entityId, command)
+        store ! command
+        timers.startSingleTimer(
+          RememberEntityTimeoutKey,
+          RememberEntityTimeout(command),
+          settings.tuningParameters.updatingStateTimeout)
+
+        context.become {
+          case RememberEntitiesShardStore.UpdateDone(entityId) =>
+            if (VerboseDebug) log.debug("Completed async write of [{}] {}", entityId, command)
+            timers.cancel(RememberEntityTimeoutKey)
+            whenDone(entityId)
+            context.become(idle)
+            unstashAll()
+          case RememberEntityTimeout(_) =>
+            throw new RuntimeException(
+              s"Async write for $entityId timed out after ${settings.tuningParameters.updatingStateTimeout.pretty}")
+
+          // below cases should handle same messages as in idle
+          case _: Terminated                           => stash()
+          case _: CoordinatorMessage                   => stash()
+          case _: ShardCommand                         => stash()
+          case _: ShardRegion.StartEntity              => stash()
+          case _: ShardRegion.StartEntityAck           => stash()
```

Comment:
maybe while you're deep into this you can remind me why a shard received a start entity ack?